daprdocs/content/en/developing-applications/building-blocks/workflow/workflow-patterns.md

```diff
@@ -1386,6 +1386,195 @@ func raiseEvent() {
+- **Partial Failure Recovery**: If a workflow fails after some steps have completed successfully, the compensation pattern allows you to undo those completed steps gracefully.
+- **Business Process Integrity**: Ensures that business processes can be properly rolled back in case of failures, maintaining the integrity of your business operations.
+- **Long-Running Processes**: For workflows that may run for hours, days, or longer, traditional locking mechanisms are impractical. Compensation provides a way to handle failures in these scenarios.
+
+Common use cases for the compensation pattern include:
+
+- **E-commerce Order Processing**: Reserve inventory, charge payment, and ship orders. If shipping fails, you need to release the inventory and refund the payment.
+- **Financial Transactions**: In a money transfer, if crediting the destination account fails, you need to rollback the debit from the source account.
+- **Resource Provisioning**: When provisioning cloud resources across multiple providers, if one step fails, you need to clean up all previously provisioned resources.
+- **Multi-Step Business Processes**: Any business process that involves multiple irreversible steps that may need to be undone in case of later failures.
+
+Dapr Workflow provides support for the compensation pattern, allowing you to register compensation activities for each step and execute them in reverse order when needed.
+
+{{< tabs Java >}}
+
+{{% codetab %}}
+<!--java-->
+
+```java
+public class PaymentProcessingWorkflow implements Workflow {
+    
+    @Override
+    public WorkflowStub create() {
+        return ctx -> {
+            ctx.getLogger().info("Starting Workflow: " + ctx.getName());
+            var orderId = ctx.getInput(String.class);
+            List<String> compensations = new ArrayList<>();
+            
+            try {
+                // Step 1: Reserve inventory
+                String reservationId = ctx.callActivity(ReserveInventoryActivity.class.getName(), orderId, String.class).await();
+                ctx.getLogger().info("Inventory reserved: {}", reservationId);
+                compensations.add("ReleaseInventory");
+                
+                // Step 2: Process payment
+                String paymentId = ctx.callActivity(ProcessPaymentActivity.class.getName(), orderId, String.class).await();
+                ctx.getLogger().info("Payment processed: {}", paymentId);
+                compensations.add("RefundPayment");
+                
+                // Step 3: Ship order
+                String shipmentId = ctx.callActivity(ShipOrderActivity.class.getName(), orderId, String.class).await();
+                ctx.getLogger().info("Order shipped: {}", shipmentId);
+                compensations.add("CancelShipment");
+                
+                // Step 4: Send confirmation
```

Comment:
I suggest to remove this activity. It's a bit confusing now if the SendConfirmation throws, then the rest is still compensated.

Reply:
I've moved the SendConfirmation to be after the try/catch so it doesn't need to be removed.